File: packages/domain/src/adapters/gates.ts
/**
 * Fail-closed adapter run gates (BB-037). No adapter runs without approved source policy.
 */
import {
  assertSourceAdapterCanCreateCandidates,
  canSourceAdapterCreateCandidates,
  type SourceKillSwitchState,
} from '../provenance/source.js';
import type { SourceRegistryEntry } from './types.js';

const RUNNABLE_REGISTRY_STATES = new Set<SourceRegistryEntry['registryState']>(['approved', 'canary']);

export function canAdapterRun(
  entry: SourceRegistryEntry,
  killSwitch?: SourceKillSwitchState | null,
): boolean {
  if (!RUNNABLE_REGISTRY_STATES.has(entry.registryState)) {
    return false;
  }
  if (!entry.approvedAt || !entry.approvedBy) {
    return false;
  }
  return canSourceAdapterCreateCandidates(entry.evidenceSource, killSwitch);
}

/**
 * Fail-closed gate: adapter runs require approved policy and an enabled evidence source.
 */
export function assertAdapterMayRun(
  entry: SourceRegistryEntry,
  killSwitch?: SourceKillSwitchState | null,
): void {
  if (!RUNNABLE_REGISTRY_STATES.has(entry.registryState)) {
    throw new Error(
      `Source adapter "${entry.contract.adapterId}" cannot run in registry state "${entry.registryState}"`,
    );
  }
  if (!entry.approvedAt || !entry.approvedBy) {
    throw new Error(
      `Source adapter "${entry.contract.adapterId}" has no approved source policy`,
    );
  }
  assertSourceAdapterCanCreateCandidates(entry.evidenceSource, killSwitch);
}

export function isCanaryMode(entry: SourceRegistryEntry): boolean {
  return entry.registryState === 'canary';
}

/**
 * Apply canary sampling: when in canary mode, only a fraction of records proceed.
 * Returns indices of records that should be processed in this run.
 */
export function selectCanaryRecordIndices(
  totalRecords: number,
  sampleFraction: number,
): readonly number[] {
  if (totalRecords <= 0) {
    return [];
  }
  if (sampleFraction <= 0 || sampleFraction > 1) {
    throw new Error('canary sampleFraction must be in (0, 1]');
  }
  const target = Math.max(1, Math.ceil(totalRecords * sampleFraction));
  const step = totalRecords / target;
  const indices: number[] = [];
  for (let i = 0; i < target; i += 1) {
    indices.push(Math.min(totalRecords - 1, Math.floor(i * step)));
  }
  return [...new Set(indices)].sort((a, b) => a - b);
}

File: workers/research/src/black_book_research/adapters/gates.py
"""Fail-closed adapter run gates mirroring domain gates (BB-037)."""

from __future__ import annotations

from .types import RUNNABLE_REGISTRY_STATES, SourceKillSwitchState, SourceRegistryEntry


def _kill_switch_blocks(
    entry: SourceRegistryEntry, kill_switch: SourceKillSwitchState | None
) -> bool:
    source = entry.evidence_source
    if not source.adapter_enabled:
        return True
    if source.kill_switch_id and kill_switch is not None:
        if kill_switch.id != source.kill_switch_id:
            raise ValueError(
                f"Kill switch id mismatch: expected {source.kill_switch_id}, got {kill_switch.id}"
            )
        if kill_switch.enabled:
            return True
    return False


def can_adapter_run(
    entry: SourceRegistryEntry, kill_switch: SourceKillSwitchState | None = None
) -> bool:
    if entry.registry_state not in RUNNABLE_REGISTRY_STATES:
        return False
    if not entry.approved_at or not entry.approved_by:
        return False
    return not _kill_switch_blocks(entry, kill_switch)


def assert_adapter_may_run(
    entry: SourceRegistryEntry, kill_switch: SourceKillSwitchState | None = None
) -> None:
    if entry.registry_state not in RUNNABLE_REGISTRY_STATES:
        raise ValueError(
            f'Source adapter "{entry.contract.adapter_id}" cannot run in registry state "{entry.registry_state}"'
        )
    if not entry.approved_at or not entry.approved_by:
        raise ValueError(
            f'Source adapter "{entry.contract.adapter_id}" has no approved source policy'
        )
    if _kill_switch_blocks(entry, kill_switch):
        raise ValueError(
            f'Source adapter "{entry.contract.adapter_id}" ({entry.id}) is disabled and cannot create candidates'
        )


def is_canary_mode(entry: SourceRegistryEntry) -> bool:
    return entry.registry_state == "canary"


def select_canary_record_indices(total_records: int, sample_fraction: float) -> tuple[int, ...]:
    if total_records <= 0:
        return ()
    if sample_fraction <= 0 or sample_fraction > 1:
        raise ValueError("canary sample_fraction must be in (0, 1]")
    target = max(1, int(total_records * sample_fraction + 0.999999))
    step = total_records / target
    indices = {min(total_records - 1, int(i * step)) for i in range(target)}
    return tuple(sorted(indices))
